interface/ceed-preconditioning.c

```diff
@@ -253,6 +254,71 @@ static int CeedOperatorCreateActivePointBlockRestriction(CeedElemRestriction rst
   return CEED_ERROR_SUCCESS;
 }
 
+/**
+  @brief Create an unsigned restriction for active operator field
+
+  @param[in]  rstr          Original CeedElemRestriction for active field
+  @param[out] unsigned_rstr Address of the variable where the newly created CeedElemRestriction will be stored
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Developer
+**/
+static int CeedOperatorCreateUnsignedRestriction(CeedElemRestriction rstr, CeedElemRestriction *unsigned_rstr) {
```

Comment:
So we're undoing the PR just merged? I'm very confused

Reply:
Yes give me a few minutes and I'll try to write up an explanation...

Sorry I did not mean to click on your name to re-request review.

Reply:
I'm curious to see. I don't love the idea of double storing the exact same indices since that's a non trivial amount of memory.

Reply:
OK here it goes. I fully agree with you so let me try to describe...

I am only realizing this after #1196 was merged, of course, but doing some more work I realized that p-multigrid will also require an unsigned element restriction based on how the dof multiplicity vector `mult_vec` is computed. Namely the scaling should be (I think) `|E| (E^T E 1)^-1` where I am abusing notation a bit: `| |` is the entry-wise absolute value and `( )^1` is a component-wise reciprocal. The issue with the `ApplyUnsigned` method is that as far as I could tell, there isn't a way to call `CeedOperatorSetField` providing a restriction and tell it to use `ApplyUnsigned` instead of `Apply`. So my thought was just to create a new restriction object sharing the `offsets` memory from the old one but without the signs. This is what I am attempting in [0724f62](https://github.com/CEED/libCEED/pull/1168/commits/0724f6211a6cbff62e9a60e24f9d1574ba274d56), yes it undoes #1196 as you noticed but would solve both cases where an unsigned element restriction is required.

In the case of `CEED_RESTRICTION_ORIENTED`, I am using `CEED_USE_POINTER` to avoid duplication of the indices. However I can't do this for the case of `CEED_RESTRICTION_CURL_ORIENTED` and so the copy of the `curl_orients` array is not nice. I thought about  creating another enum entry for `CEED_RESTRICTION_UNSIGNED_CURL_ORIENTED` which would allow us to avoid the memory duplication and then just applying the `abs` inside of `ElemRestrictionApply` for that restriction type. I wasn't sure if that is a better idea so was going to ask.

Anyway, maybe I am missing something and we can use the `ApplyUnsigned` approach of #1196 even for the multiplicity issue, so I'm curious what you and others think here. Thank you!

Reply:
If the issue is with multiplicity, we should target that directly. Creating restrictions on the GPU without extra memcopies and allocations is going to be harder than on the CPU.

We could modify the signature of `CeedOperatorMultigridLevelCreate*` to take an optional `rstr_p_mult_fine`.

Another option - We could instead add a `CeedElemRestrictionCreateUnsigned(rstr_old, &rstr_unsigned)` that uses the same backend data but swaps `Apply` to be `ApplyUnsigned`. That might be the cleanest approach. There are some details to iron out, but I could dummy up a version tomorrow morning pretty quickly.

Reply:
> Another option - We could instead add a `CeedElemRestrictionCreateUnsigned(rstr_old, &rstr_unsigned)` that uses the same backend data but swaps `Apply` to be `ApplyUnsigned`. That might be the cleanest approach. There are some details to iron out, but I could dummy up a version tomorrow morning pretty quickly.

This seems like a nice way of solving this, I agree! Would be very helpful if you can come up with a demo implementation of this.

Reply:
This? https://github.com/CEED/libCEED/pull/1206

Reply:
Sorry for the delay but this looks good. I'm trying to straighten out if this multiplicity calculation works or what it should look like in the case of the changes provided in this branch, and it isn't super straightforward to me. If I think of `\mathcal{E} = TE` where `E` is the standard restriction of L-dofs to E-dofs and `T` is the block 2x2 matrix with orientations and transformations like `[-1 0; -1 1]` (for example), then I don't know if `|\mathcal{E}| (|\mathcal{E}|^T|\mathcal{E}| 1)^-1` is actually the right multiplicity to use to get an interpolating operator. It's almost like you always want to use only `E (E^T E 1)^-1` and ignore the auxiliary orientations and transformations data altogether. Though I'm also not sure about this.

Reply:
The H^Div and H^Curl interpolation operations are a bit of a mystery to me currently, so I'm not quite sure.

The multiplicity should scale for duplication across element and process boundaries, but I'm not sure what that is for those bases

Reply:
Resolving for now due to the discussion [here](https://github.com/CEED/libCEED/pull/1206#discussion_r1184507683). The case of curl-oriented element restrictions is not accounted for yet but will be left to resolve in the future outside the scope of this PR.